src/multipart/multipart.rs

```diff
@@ -0,0 +1,469 @@
+                    // each header once.
+                    let header_length = field.header().len();
+                    // The additions mimick the format string out of which the field is constructed
+                    // in RequestReader. Not the cleanest solution because if that format string is
+                    // ever changed then this formula needs to be changed too which is not an
+                    // obvious dependency in the code.
+                    length += 2 + self.boundary.len() as u64 + 2 + header_length as u64 + 4 + value_length + 2
+                }
+                _ => return None,
+            }
+        }
+        // If there is a at least one field there is a special boundary for the very last field.
+        if self.fields.len() != 0 {
+            length += 2 + self.boundary.len() as u64 + 2
+        }
+        Some(length)
+    }
+}
+
+/// A field in a multipart request.
+pub struct MultipartField {
+    name: Cow<'static, str>,
+    value: Box<Read + Send>,
+    value_length: Option<u64>,
+    mime: Option<Mime>,
+    file_name: Option<Cow<'static, str>>,
+}
+
+// TODO: MultipartField cannot derive debug because value is not Debug
+// Not sure how to best resolve this...
+impl std::fmt::Debug for MultipartField {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "")
+    }
+}
+
+impl MultipartField {
+    /// Makes a String parameter.
+    ///
+    /// ```
+    /// reqwest::MultipartField::param("key", "value");
+    /// ```
+    ///
+    /// ```
+    /// let key: String = "key".to_string();
+    /// let value: String = "value".to_string();
+    /// reqwest::MultipartField::param(key, value);
+    /// ```
+    ///
+    pub fn param<T: Into<Cow<'static, str>>, U: AsRef<[u8]> + Send + 'static>(name: T, value: U) -> MultipartField {
+        let value_length = Some(value.as_ref().len() as u64);
+        MultipartField {
+            name: name.into(),
+            value: Box::new(std::io::Cursor::new(value)),
+            value_length: value_length,
+            mime: None,
+            file_name: None,
+        }
+    }
+    /// Adds a generic reader.
+    ///
+    /// ```
+    /// use std::io::empty;
+    /// let reader = empty();
+    /// reqwest::MultipartField::reader("key", reader);
+    /// ```
+    ///
+    pub fn reader<T: Into<Cow<'static, str>>, U: Read + Send + 'static>(name: T, value: U) -> MultipartField {
+        MultipartField {
+            name: name.into(),
+            value: Box::from(value),
+            value_length: None,
+            mime: None,
+            file_name: None,
+        }
+    }
+    /// Makes a file parameter.
+    /// Defaults to mime type application/octet-stream.
+    ///
+    /// ```no_run
+    /// reqwest::MultipartField::file("key", "/path/to/file");
+    /// ```
+    ///
+    /// # Errors
+    /// Errors when the file cannot be opened.
+    pub fn file<T: Into<Cow<'static, str>>, U: AsRef<std::path::Path>>(
```

Comment:
People who already have a `File` open can pass that via `reader`. I have added a doc example for that.